### PR TITLE
Change logic Github Action for new CI/CD flow

### PR DESCRIPTION
**Pull request by harris1111**
## Type

- [ ] Feature
- [ ] Bug fix
- [x] CI/CD
- [ ] Documentation
- [ ] Other

## Description

Changes logic for Github action to detect changes between 2 last merge instead of two last commit. Also, I use an unified step to build and push changed packages instead of separating it. 
Moreover, I add a step to make changes and push to GitOps repo for faster deployment in test environment.

### CODE REVIEW

File: .github/workflows/build-and-push-docker.yml [hunk 1]
```diff
@@ -105,53 +75,72 @@ jobs:
...
+            else
+              echo "$SERVICE not changed or invalid Docker repo, skipping."
+            fi
+          done
+          echo "::set-output name=changed_services::$CHANGED_SERVICES"
+
+  gitops-push:
+    name: Push Changes to GitOps
+    needs: build
+    runs-on: ubuntu-latest
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:117]:
```suggestion
    [self-hosted, linux]
```

File: .github/workflows/build-and-push-docker.yml [hunk 2]
```diff
@@ -5,93 +5,63 @@ on:
...
-  DOCKER_REPO_BROKER_SERVICE: orochinetwork/zkdb-broker-service
-  DOCKER_REPO_PROOF_SERVICE: orochinetwork/zkdb-proof-service
-  DOCKER_REPO_SERVERLESS: orochinetwork/zkdb-serverless
-  NPM_ACCESS_TOKEN: ${{ secrets.NPM_ACCESS_TOKEN }}
-
 jobs:
   detect-changes:
     name: Detect Changes
-    runs-on: [self-hosted, linux]
+    runs-on: ["self-hosted", "linux", "ubuntu-latest"]
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:11]:
```suggestion
    runs-on: [self-hosted, linux]
```

File: .github/workflows/build-and-push-docker.yml [hunk 3]
```diff
@@ -105,53 +75,72 @@ jobs:
...
+          done
+          echo "::set-output name=changed_services::$CHANGED_SERVICES"
+
+  gitops-push:
+    name: Push Changes to GitOps
+    needs: build
+    runs-on: ubuntu-latest
+    env:
+      version: ${{ github.sha }}
+      gitopsbranch: devops/test-gitops
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:120]: Why we need this branch?

File: .github/workflows/build-and-push-docker.yml [hunk 4]
```diff
@@ -105,53 +75,72 @@ jobs:
...
-            docker push ${{ env.DOCKER_REPO_SERVERLESS }}:${{ github.sha }}
-          else
-            echo "Serverless Service not in changed packages."
-          fi
+          eval $(ssh-agent -s)
+          echo "${{ secrets.SSH_KEY }}" | tr -d '\r' | ssh-add -
+          mkdir -p ~/.ssh
+          ssh-keyscan github.com >> ~/.ssh/known_hosts
+
+          git config --global user.name "GitOps Updater"
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:129]: pls use github secret instead

File: .github/workflows/build-and-push-docker.yml [hunk 5]
```diff
@@ -105,53 +75,72 @@ jobs:
...
+          git config --global user.email "minhan112001@gmail.com"
+          git clone git@github.com:ZKP-Labs/test-argoCD.git
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:131]: use github secret env

File: .github/workflows/build-and-push-docker.yml [hunk 6]
```diff
@@ -105,53 +75,72 @@ jobs:
...
-          else
-            echo "Serverless Service not in changed packages."
-          fi
+          eval $(ssh-agent -s)
+          echo "${{ secrets.SSH_KEY }}" | tr -d '\r' | ssh-add -
+          mkdir -p ~/.ssh
+          ssh-keyscan github.com >> ~/.ssh/known_hosts
+
+          git config --global user.name "GitOps Updater"
+          git config --global user.email "minhan112001@gmail.com"
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:130]: use github secret env

File: .github/workflows/build-and-push-docker.yml (hunk 3 above)
Comment (harris1111) [reply .github/workflows/build-and-push-docker.yml:120]: This branch is used in `test-argoCD` repo for argoCD to auto apply new changes. The purpose creating this is when we have new commit in main at `zkDatabase`, the commit ID will also be the tag of the image, and this commitID will be push in to this branch, so we keep the changes in a separate branch, not affecting other branch in `test-argoCD` repo. If we have other modification in helm chart, we simply merge these changes into this branch. 
Moreover, it's only a test branch, I will discuss with you about its name when my PR is merged.

Comment (dqtkien) [reply .github/workflows/build-and-push-docker.yml:120]: I think we should stick with `main` branch.

File: .github/workflows/build-and-push-docker.yml [hunk 7]
```diff
@@ -105,53 +75,74 @@ jobs:
...
-            echo "Serverless Service not in changed packages."
-          fi
+          eval $(ssh-agent -s)
+          echo "${{ secrets.SSH_KEY }}" | tr -d '\r' | ssh-add -
+          mkdir -p ~/.ssh
+          ssh-keyscan github.com >> ~/.ssh/known_hosts
+
+          git config --global user.name "${{ secrets.GITOPS_USER }}"
+          git config --global user.email "${{ secrets.GITOPS_EMAIL }}"
+          git clone git@github.com:${{ secrets.GITOPS_REPO }}
```
Comment (dqtkien) [inline .github/workflows/build-and-push-docker.yml:131]:
```suggestion
          git clone ${{ secrets.GITOPS_REPO }}
```

**Review by dqtkien — APPROVED** [review]

**Pull request merged**